File: sdks/go/approvals.go
package atap

import "context"

// ApprovalAPI provides approval creation, response, listing, and revocation.
type ApprovalAPI struct {
	client *Client
}

// Create creates an approval request. Requires atap:send scope.
func (a *ApprovalAPI) Create(ctx context.Context, fromDID, toDID string, subject ApprovalSubject, via string) (*Approval, error) {
	body := map[string]interface{}{
		"from": fromDID,
		"to":   toDID,
		"subject": map[string]interface{}{
			"type":    subject.Type,
			"label":   subject.Label,
			"payload": subject.Payload,
		},
	}
	if via != "" {
		body["via"] = via
	}

	data, err := a.client.authedRequest(ctx, "POST", "/v1/approvals", &RequestOptions{JSONBody: body})
	if err != nil {
		return nil, err
	}
	return parseApproval(data), nil
}

// Respond responds to an approval (approve). Requires atap:send scope.
func (a *ApprovalAPI) Respond(ctx context.Context, approvalID, signature string) (*Approval, error) {
	body := map[string]interface{}{
		"signature": signature,
	}
	data, err := a.client.authedRequest(ctx, "POST", "/v1/approvals/"+approvalID+"/respond", &RequestOptions{JSONBody: body})
	if err != nil {
		return nil, err
	}
	return parseApproval(data), nil
}

// List lists approvals addressed to the authenticated entity. Requires atap:inbox scope.
func (a *ApprovalAPI) List(ctx context.Context) ([]Approval, error) {
	data, err := a.client.authedRequest(ctx, "GET", "/v1/approvals", nil)
	if err != nil {
		return nil, err
	}
	return parseApprovalList(data), nil
}

// Revoke revokes an approval. Requires atap:revoke scope.
func (a *ApprovalAPI) Revoke(ctx context.Context, approvalID string) (*Approval, error) {
	data, err := a.client.authedRequest(ctx, "DELETE", "/v1/approvals/"+approvalID, nil)
	if err != nil {
		return nil, err
	}
	return parseApproval(data), nil
}

func parseApproval(data map[string]interface{}) *Approval {
	approval := &Approval{
		ID:          getString(data, "id"),
		State:       getString(data, "state"),
		CreatedAt:   getString(data, "created_at"),
		ValidUntil:  getString(data, "valid_until"),
		FromDID:     getString(data, "from"),
		ToDID:       getString(data, "to"),
		Via:         getString(data, "via"),
		Parent:      getString(data, "parent"),
		TemplateURL: getString(data, "template_url"),
		RespondedAt: getString(data, "responded_at"),
	}

	if sigs, ok := data["signatures"].(map[string]interface{}); ok {
		approval.Signatures = make(map[string]string)
		for k, v := range sigs {
			if s, ok := v.(string); ok {
				approval.Signatures[k] = s
			}
		}
	}

	if fanOut, ok := data["fan_out"].(float64); ok {
		n := int(fanOut)
		approval.FanOut = &n
	}

	if subj, ok := data["subject"].(map[string]interface{}); ok {
		s := &ApprovalSubject{
			Type:  getString(subj, "type"),
			Label: getString(subj, "label"),
		}
		if r, ok := subj["reversible"].(bool); ok {
			s.Reversible = r
		}
		if p, ok := subj["payload"].(map[string]interface{}); ok {
			s.Payload = p
		}
		approval.Subject = s
	}

	return approval
}

func parseApprovalList(data map[string]interface{}) []Approval {
	// Handle array-like responses via items or approvals key.
	var items []interface{}
	if arr, ok := data["approvals"].([]interface{}); ok {
		items = arr
	} else if arr, ok := data["items"].([]interface{}); ok {
		items = arr
	}

	result := make([]Approval, 0, len(items))
	for _, item := range items {
		if m, ok := item.(map[string]interface{}); ok {
			result = append(result, *parseApproval(m))
		}
	}
	return result
}

File: sdks/javascript/src/approvals.ts
/**
 * Approval operations.
 */

import type { ATAPClient } from "./client.js";
import type { Approval, ApprovalSubject } from "./models.js";

/** Create, respond to, list, and revoke approvals. */
export class ApprovalAPI {
  private readonly _client: ATAPClient;

  constructor(client: ATAPClient) {
    this._client = client;
  }

  /**
   * Create an approval request. Requires atap:send scope.
   *
   * @param fromDid - Requester DID.
   * @param toDid - Approver DID (or org DID for fan-out).
   * @param subject - The approval subject with type, label, and payload.
   * @param options.via - Optional mediating system DID.
   */
  async create(
    fromDid: string,
    toDid: string,
    subject: ApprovalSubject,
    options: { via?: string } = {},
  ): Promise<Approval> {
    const body: Record<string, unknown> = {
      from: fromDid,
      to: toDid,
      subject: {
        type: subject.type,
        label: subject.label,
        payload: subject.payload || {},
      },
    };
    if (options.via) body.via = options.via;

    const data = await this._client._authedRequest("POST", "/v1/approvals", {
      jsonBody: body,
    });
    return parseApproval(data);
  }

  /**
   * Respond to an approval (approve). Requires atap:send scope.
   *
   * @param approvalId - The approval ID (apr_...).
   * @param signature - JWS signature from the approver.
   */
  async respond(approvalId: string, signature: string): Promise<Approval> {
    const data = await this._client._authedRequest(
      "POST",
      `/v1/approvals/${approvalId}/respond`,
      { jsonBody: { signature } },
    );
    return parseApproval(data);
  }

  /** List approvals addressed to the authenticated entity. Requires atap:inbox scope. */
  async list(): Promise<Approval[]> {
    const data = await this._client._authedRequest("GET", "/v1/approvals");
    if (Array.isArray(data)) {
      return (data as Record<string, unknown>[]).map(parseApproval);
    }
    const items =
      (data.approvals as Record<string, unknown>[]) ||
      (data.items as Record<string, unknown>[]) ||
      [];
    return items.map(parseApproval);
  }

  /** Revoke an approval. Requires atap:revoke scope. */
  async revoke(approvalId: string): Promise<Approval> {
    const data = await this._client._authedRequest(
      "DELETE",
      `/v1/approvals/${approvalId}`,
    );
    return parseApproval(data);
  }
}

export function parseApproval(data: Record<string, unknown>): Approval {
  let subject: ApprovalSubject | undefined;
  if (data.subject) {
    const s = data.subject as Record<string, unknown>;
    subject = {
      type: (s.type as string) || "",
      label: (s.label as string) || "",
      reversible: (s.reversible as boolean) || false,
      payload: s.payload as Record<string, unknown> | undefined,
    };
  }
  return {
    id: (data.id as string) || "",
    state: (data.state as string) || "",
    createdAt: (data.created_at as string) || "",
    validUntil: data.valid_until as string | undefined,
    fromDid: (data.from as string) || "",
    toDid: (data.to as string) || "",
    via: (data.via as string) || "",
    parent: (data.parent as string) || "",
    subject,
    templateUrl: (data.template_url as string) || "",
    signatures: (data.signatures as Record<string, string>) || {},
    respondedAt: data.responded_at as string | undefined,
    fanOut: data.fan_out as number | undefined,
  };
}
